SPEC.md

```diff
@@ -0,0 +1,121 @@
+This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.
+
+## Core Language
+A CNI document consists of an arbitrary amount of statements.
+A statement may either be a section declaration, key-value pair, or key-rawvalue pair.
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
+
+A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
+If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.
+
+Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
+
+### Flexspace
+In implementations with the "flexspace" extension, whitespace within statements is not significant.
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
+
```

Comment:
No, comments may not be placed within such constructs, since comments are statements and not whitespace.
This extension is already mostly there since it's harder to disallow it than it is to let it be.
Having comments inside of flexspace may be an extension, but I don't actually expect people to ever do this as it is.